src/wp-includes/rest-api.php

```diff
@@ -295,6 +296,25 @@ function rest_api_loaded() {
 	 */
 	define( 'REST_REQUEST', true );
 
+	// Populate the correct $_SERVER variables via an alternate header for fastcgi compatibility.
+	if ( isset( $_SERVER['HTTP_WP_AUTHORIZATION'] ) && preg_match( '%^Basic [a-z\d/+]*={0,2}$%i', $_SERVER['HTTP_WP_AUTHORIZATION'] ) ) {
+		// Removing `Basic ` the token would start six characters in.
+		$token = substr( $_SERVER['HTTP_WP_AUTHORIZATION'], 6 );
+		$userpass = base64_decode( $token );
+		list( $_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW'] ) = explode( ':', $userpass );
+	}
+
+	// Determine the user that the request should be run under (if any).
+	if ( isset( $_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW'] ) ) {
+		// We are explicitly only accepting HTTP Basic Auth for HTTPS requests.
+		if ( ! is_ssl() ) {
+			wp_send_json_error( __( 'HTTP Basic Auth is unavailable for non-HTTPS requests.' ), 403 );
+			die();
+		}
+		$GLOBALS['user'] = wp_authenticate( $_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW'] );
+		wp_set_current_user( $GLOBALS['user']->id );
```

Comment:
So you shouldn't need to use the globals any longer for user -- just set it to a local `$user` variable, and then pass that to wp_set_current_user() if it's a WP_User object.

https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/pluggable.php#L38

You don't even need to pick the ID out of it, as wp_set_current_user instantiates a WP_User object with the first arg passed in, and WP_User's constructor will accept a WP_User as its first argument.

https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/class-wp-user.php#L134-L136

Reply:
Thanks. I updated accordingly and it's cleaner now :)